### PR TITLE
add partial versioning support

### PR DESCRIPTION
**Pull request by wrightp**
Signed-off-by: Patrick Wright <patrick@chef.io>

This was was added for a couple reasons:
1) partial versioning support is required to support the test-kitchen config deprecation project. In this case we are maintaining feature parity with the `require_chef_omnibus` options.
2) It's cool and probably helpful to a lot of people

`product_version` now supports versions like `12`, `12.1`, `11`, `11.1`

This change also uncovered a version sorting error. We have been lucky not to encounter because of the consistency in the promotion of artifacts. However, there were inconsistencies with older versions (11). Versions are now sorted using mixlib-versioning.

See #192

@chef/engineering-services

### CODE REVIEW

File: lib/mixlib/install/options.rb [hunk 1]
```diff
@@ -110,6 +110,10 @@ def latest_version?
         product_version.to_sym == :latest
       end
 
+      def partial_version?
+        !Mixlib::Versioning.parse(product_version) && !latest_version?
```
Comment (wrightp) [inline lib/mixlib/install/options.rb]: This is pretty loose. I could do a simple reg exp here. Opinions?

**Comment by wrightp** [comment]
```
bundle exec mixlib-install download chef -v 11 -p ubuntu -l 14.04 --url --attributes
https://packages.chef.io/files/stable/chef/11.18.12/ubuntu/14.04/chef_11.18.12-1_amd64.deb
{
  "architecture": "x86_64",
  "license": null,
  "license_content": null,
  "md5": "5068a866a833170bb99baa4813abe4e3",
  "platform": "ubuntu",
  "platform_version": "14.04",
  "product_description": "Chef Client",
  "product_name": "chef",
  "sha1": "c504957ff6721c15aae8c3741a5418d222c3626e",
  "sha256": "2a59c3cd6818f28ba1d946b390d97ce8400b46e7f5c6362cd6202b5e961b5987",
  "software_dependencies": null,
  "url": "https://packages.chef.io/files/stable/chef/11.18.12/ubuntu/14.04/chef_11.18.12-1_amd64.deb",
  "version": "11.18.12"
}
```

**Comment by wrightp** [comment]
```
bundle exec mixlib-install download chef -c current -v 13 -p ubuntu -l 14.04 --url --attributes
https://packages.chef.io/files/current/chef/13.0.99/ubuntu/14.04/chef_13.0.99-1_amd64.deb
{
  "architecture": "x86_64",
  "license": "Apache-2.0",
  "license_content": null,
  "md5": "e120f5a5ef262bec39b80f5bc5c6dc8d",
  "platform": "ubuntu",
  "platform_version": "14.04",
  "product_description": "Chef Client",
  "product_name": "chef",
  "sha1": "9255ad50116b3b80db6a6c8ba6cab8f71cb4b02e",
  "sha256": "ff3d341be7b522cbf0a0423aa771c1efe3dd94e403c10f0242f75e45b1210c7f",
  "software_dependencies": null,
  "url": "https://packages.chef.io/files/current/chef/13.0.99/ubuntu/14.04/chef_13.0.99-1_amd64.deb",
  "version": "13.0.99"
}
```

File: lib/mixlib/install/options.rb (hunk 1 above)
Comment (scotthain) [inline lib/mixlib/install/options.rb]: Is there a speed difference?

Comment (wrightp) [inline lib/mixlib/install/options.rb]: Yes, but this method only gets executed once or twice when calling `#artifact_info`

Comment (scotthain) [inline lib/mixlib/install/options.rb]: 👍 I don't feel strongly about it :)

Comment (schisamo) [inline lib/mixlib/install/options.rb]: I think this approach is good as it yields the actual parsing down to mixlib-versioning...which means we don't maintain this parsing logic in two separate spots. You might want to flip your AND check and do `!latest_version?` first as that would save parsing the version if it is the `:latest`.

**Pull request merged**